crates/gpui/src/platform/linux/x11/client.rs

```diff
@@ -1735,3 +1763,231 @@ fn xdnd_send_status(
+                    .map(|data| scroll_data_to_axis_state(data, old_vertical));
+                if horizontal.is_none() && vertical.is_none() {
+                    None
+                } else {
+                    Some((
+                        info.deviceid,
+                        PointerDeviceState {
+                            horizontal: horizontal.unwrap_or_else(Default::default),
+                            vertical: vertical.unwrap_or_else(Default::default),
+                        },
+                    ))
+                }
+            }),
+    );
+    if pointer_device_states.is_empty() {
+        log::error!("Found no xinput mouse pointers.");
+    }
+    return pointer_device_states;
+}
+
+/// Returns true if the device is a pointer device. Does not include pointer device groups.
+fn is_pointer_device(type_: xinput::DeviceType) -> bool {
+    type_ == xinput::DeviceType::SLAVE_POINTER
+}
+
+fn scroll_data_to_axis_state(
+    data: &xinput::DeviceClassDataScroll,
+    old_axis_state_with_valid_scroll_value: Option<&ScrollAxisState>,
+) -> ScrollAxisState {
+    ScrollAxisState {
+        valuator_number: Some(data.number),
+        multiplier: SCROLL_LINES / fp3232_to_f32(data.increment),
+        scroll_value: old_axis_state_with_valid_scroll_value.and_then(|state| state.scroll_value),
+    }
+}
+
+fn reset_all_pointer_device_scroll_positions(
+    pointer_device_states: &mut BTreeMap<xinput::DeviceId, PointerDeviceState>,
+) {
+    pointer_device_states
+        .iter_mut()
+        .for_each(|(_, device_state)| reset_pointer_device_scroll_positions(device_state));
+}
+
+fn reset_pointer_device_scroll_positions(pointer: &mut PointerDeviceState) {
+    pointer.horizontal.scroll_value = None;
+    pointer.vertical.scroll_value = None;
+}
+
+/// Returns the scroll delta for a smooth scrolling motion event, or `None` if no scroll data is present.
+fn get_scroll_delta_and_update_state(
+    pointer: &mut PointerDeviceState,
+    event: &xinput::MotionEvent,
+    modifiers: &Modifiers,
+) -> Option<Point<f32>> {
+    let horizontal_axis_index = pointer
+        .horizontal
+        .valuator_number
+        .and_then(|valuator_number| get_valuator_axis_index(&event.valuator_mask, valuator_number));
+    let vertical_axis_index = pointer
+        .vertical
+        .valuator_number
+        .and_then(|valuator_number| get_valuator_axis_index(&event.valuator_mask, valuator_number));
+
+    let mut delta_x = None;
+    let mut delta_y = None;
+
+    if let Some(horizontal_axis_index) = horizontal_axis_index {
+        if let Some(value) = event.axisvalues.get(horizontal_axis_index) {
+            delta_x = Some(get_axis_scroll_delta_and_update_state(
+                &mut pointer.horizontal,
+                fp3232_to_f32(*value),
+            ));
+        } else {
+            log::error!(
+                "Encountered invalid XInput valuator_mask, scrolling may not work properly."
+            )
+        }
+    }
```

Comment:
If the `?` change above is made, this can be simplified a lot by dropping the first condition. And since we always require a `delta_x`, that second `if let Some(value) = ` can also be turned into `let delta_x_value = event.axisvalues.get(horizontal_axis_index)?;` I think.

Reply:
Good idea, done!